src/par_iter/internal.rs

```diff
@@ -16,7 +16,12 @@ pub trait ProducerCallback<ITEM> {
 /// A producer which will produce a fixed number of items N. This is
 /// not queryable through the API; the consumer is expected to track
 /// it.
-pub trait Producer: IntoIterator + Send + Sized {
+pub trait Producer: Send + Sized
+{
+    type Item;
+    type IntoIter: Iterator<Item = Self::Item> + DoubleEndedIterator + ExactSizeIterator;
+    fn into_iter(self) -> Self::IntoIter;
```

Comment:
Yes, that's where I got the idea from. Otherwise I probably would have done the same as you and spent a long time diagnosing why the compiler wasn't propagating the constraint.